activerecord/lib/active_record/connection_handling.rb

```diff
@@ -333,6 +347,14 @@ def clear_cache! # :nodoc:
       connection_pool.schema_cache.clear!
     end
 
+    def shard_keys
+      connection_class_for_self.instance_variable_get(:@shard_keys) || []
```

Comment:
Why `instance_variable_get` and not calling a reader method? I'd prefer we always set the IV (even if it's empty) and call a method.

Reply:
To be honest, I had a hard time deciding if `instance_variable_get` or a class variable (with `cattr_reader`) was a better approach.

I went with `instance_variable_get` because I felt like it adds clarity about exactly where `@shard_keys` is set. 

A contrived example:

```ruby
class Animals < ActiveRecord::Base
  self.abstract_class = true
  
  connects_to shards: {
    shard_one: { writing: :animal_shard_one },
    shard_two: { writing: :animal_shard_two },
  }
  
  @shard_keys = [:shard_one, :shard_two]
end

class Birds < Animals
end

class Robin < Birds
end
```

Calling `Robin.connection_class_for_self.instance_variable_get(:@shard_keys)` provides a hint that the values should live on `Animals` and not elsewhere.

A class variable could serve the same purpose but I kinda like the approach of "Hey, model, ask your connection class what shards it can connect to."

I hope my thought process there made any sense! 😅 I in no way feel strongly about the implementation though and defer to your preferences. 🙂

We could do away with `|| []` by setting `@shard_keys = []` on `ActiveRecord::Base` if that's preferred!

Reply:
You could define the attribute reader on the class instance : 

```ruby
class Animals < ActiveRecord::Base
  self.abstract_class = true
  class << self
    attr_reader :shard_keys
  end
  
  connects_to shards: {
    shard_one: { writing: :animal_shard_one },
    shard_two: { writing: :animal_shard_two },
  }
  
  @shard_keys = [:shard_one, :shard_two]
end

class Birds < Animals
end

class Robin < Birds
end

Animals.shard_keys #=> [:shard_one, :shard_two]
Robin.shard_keys #=> nil
```

The downside is that child classes will also inherit from reader, but theirs will return nil. 
You could raise a `NoMethodError` inside the reader if `@shard_keys` is nil, but it starts to be complicated for a fairly small gain in clarity.

Reply:
Yeah, the other downside with that approach is that, with the current implementation, `Robin.sharded?` would return `false`, even though it's a model that connects to shards.

Reply:
True, we could changed `.sharded?` to call `connection_class_for_self.shard_keys.any?` with the above implementation. 

I think with problem with @JoeDupuis's example is that `Robin.shard_keys` returning `nil` would be unexpected because its connection class connects to shards.

That said, there's probably an elegant pattern we could use as an alternative that I'm not thinking of. 🤔

Reply:
> I think with problem with @JoeDupuis's example is that `Robin.shard_keys` returning `nil` would be unexpected because its connection class connects to shards.

I guess what I was getting at is that I don't see how this _isn't_ a problem in the current PR. If I am wrong I just want to make sure we have a regression test in case I want to refactor later. 😄

Reply:
I think tests provide the coverage:

```ruby
    def test_connects_to_sets_shard_keys
      assert_empty(ActiveRecord::Base.shard_keys)
      assert_equal([:shard_one, :shard_two], ShardedBase.shard_keys)
    end

    def test_connects_to_sets_shard_keys_for_descendents
      assert_equal(ShardedBase.shard_keys, ShardedModel.shard_keys)
    end

    def test_sharded?
      assert_not ActiveRecord::Base.sharded?
      assert_not UnshardedBase.sharded?
      assert_not UnshardedModel.sharded?

      assert_predicate ShardedBase, :sharded?
      assert_predicate ShardedModel, :sharded?
    end
```

We call `.connected_to` on `ShardedBase`. `ShardedModel` inherits from `ShardedBase` and thus they should share the same shard keys:

```ruby
  assert_equal(ShardedBase.shard_keys, ShardedModel.shard_keys)
```
@JoeDupuis's example sets an `attr_reader` on the connection class but that causes subclasses to be unaware if they can connect to shards or not.

I hope I'm understanding the question correctly! Let me know!

Reply:
Great, yes. Thank you!